hold references to storages during TorchScript serializaiton

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59642 hold references to storages during TorchScript serializaiton**

Uses StorageContext to hold a reference to all storages seen during TorchScript serialization to allow for tensors to be created/destroyed during serialization process. Tracking of the storages solves for the ABA memory problem.

Differential Revision: [D28968947](https://our.internmc.facebook.com/intern/diff/D28968947)

Commit:
hold references to storages during TorchScript serializaiton

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28968947).

Commit:
Update on "hold references to storages during TorchScript serializaiton"


Differential Revision: [D28968947](https://our.internmc.facebook.com/intern/diff/D28968947)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28968947).

cccclai commented:
Can this pr be cherry-picked? Otherwise heavy save/load process could be flaky

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -475,10 +475,11 @@ void ScriptModuleSerializer::writeArchive(
-              ".storage");
+                  tensor.storage().unsafeGetStorageImpl()));
+          tensor_names.push_back(string_id + ".storage");
+          storage_context_.addStorage(string_id, tensor.storage());
```
I feel it's a great idea to use StorageContext. The storage_context_ lives in the life span of the serializer so it should be safe for any tensor serialization. @cccclai could you test if it works for bytecode v5 serialization with quantized models? It looks different than #59488 that you've already tested.

cccclai replied:
Test it by re-introduce #58641 and use the following code to test:

```
for i in range(20):
    print(i)
    m = torch.jit.load(model_path)
    m._save_for_lite_interpreter(model_resave_path)
    mm = _load_for_lite_interpreter(model_resave_path)
```
![image](https://user-images.githubusercontent.com/16430979/121267234-bf852b00-c870-11eb-9dd4-8ca7cbf96c8e.png)

iseeyuan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -475,10 +475,11 @@ void ScriptModuleSerializer::writeArchive(
               std::to_string(reinterpret_cast<std::intptr_t>(
-                  tensor.storage().unsafeGetStorageImpl())) +
-              ".storage");
+                  tensor.storage().unsafeGetStorageImpl()));
```
I guess the only caveat of using StorageImpl is that the id may change from run to run. Should not be a real issue, but for metadata comparison on data.pkl or bytecode.pkl, the tensor metadata may be different literally, although the tensor content does not change.

Lilyjjo replied:
Yes, the loss of determinism in the naming is not the best. Others were wanting this to be fixed as well. I can address it in a follow up PR

cccclai reviewed (approved):
LGTM from mobile side.

Lilyjjo commented:
Do we want to try to get this PR into the 1.9 release? I that people were still adding to the branch earlier today https://github.com/pytorch/pytorch/issues/58518

cccclai commented:
> Do we want to try to get this PR into the 1.9 release? I saw that people were still adding to the branch earlier today #58518
> 
> It would fix the serialization issues I think that are currently present in the release

Let's try. Can you comment under the tracker? I will also comment there as well

iseeyuan reviewed (approved):
LGTM. Thanks!
